src/internet/radiobrowser/radiobrowserservice.cpp

```diff
@@ -0,0 +1,390 @@
+
+bool operator<(const RadioBrowserService::Stream& a,
+               const RadioBrowserService::Stream& b) {
+  return a.name_.compare(b.name_, Qt::CaseInsensitive) < 0;
+}
+
+const char* RadioBrowserService::kServiceName = "Radio-Browser.info";
+const char* RadioBrowserService::kSettingsGroup = "RadioBrowser";
+const char* RadioBrowserService::kSchemeName = "radiobrowser";
+
+QString RadioBrowserService::SearchUrl = "%1/json/stations/byname/%2?limit=%3";
+QString RadioBrowserService::PlayClickUrl = "%1/json/url/%2";
+
+QList<RadioBrowserService::Branch> RadioBrowserService::BranchList = {
+    {"By Country", "%1/json/countries",
+     "%1/json/stations/bycountryexact/%2?hidebroken=true", Type_Category},
+    {"By Language", "%1/json/languages",
+     "%1/json/stations/bylanguageexact/%2?hidebroken=true", Type_Category},
+    {"By Tag", "%1/json/tags", "%1/json/stations/bytagexact/%2?hidebroken=true",
+     Type_Category},
+    {"By Codec", "%1/json/codecs",
+     "%1/json/stations/bycodecexact/%2?hidebroken=true", Type_Category},
+    {"Top 100 Clicked", "",
+     "%1/json/stations/search?order=clickcount&reverse=true&limit=100",
+     Type_Top100},
+    {"Top 100 Voted", "",
+     "%1/json/stations/search?order=votes&reverse=true&limit=100", Type_Top100},
+    {"Top 100 Trending", "",
+     "%1/json/stations/search?order=clicktrend&reverse=true&limit=100",
+     Type_Top100}};
+
+RadioBrowserService::RadioBrowserService(Application* app,
+                                         InternetModel* parent)
+    : InternetService(kServiceName, app, parent, parent),
+      root_(nullptr),
+      context_menu_(nullptr),
+      network_(new NetworkAccessManager(this)),
+      name_(kServiceName),
+      url_scheme_(kSchemeName),
+      url_handler_(new RadioBrowserUrlHandler(app, this, this)),
+      homepage_url_(QUrl("https://www.radio-browser.info")),
+      icon_(IconLoader::Load("radiobrowser", IconLoader::Provider)) {
+  app_->player()->RegisterUrlHandler(url_handler_);
+  app_->global_search()->AddProvider(
+      new RadioBrowserSearchProvider(app_, this, this));
+}
+
+RadioBrowserService::~RadioBrowserService() { delete context_menu_; }
+
+QStandardItem* RadioBrowserService::CreateRootItem() {
+  root_ = new QStandardItem(icon_, name_);
+  root_->setData(true, InternetModel::Role_CanLazyLoad);
+  return root_;
+}
+
+void RadioBrowserService::LazyPopulate(QStandardItem* item) {
+  switch (item->data(InternetModel::Role_Type).toInt()) {
+    case InternetModel::Type_Service:
+      RefreshRootItem();
+      break;
+    case RadioBrowserService::Type_Category:
+      RefreshCategory(item);
+      break;
+    case RadioBrowserService::Type_CategoryItem:
+      RefreshCategoryItem(item);
+      break;
+    case RadioBrowserService::Type_Top100:
+      RefreshTop100(item);
+      break;
+    default:
+      break;
+  }
+}
+
+void RadioBrowserService::RefreshRootItem() {
+  if (!EnsureServerConfig()) {
+    ShowConfig();
+    return;
+  }
```

Comment:
Since there is a default server and credentials aren't required, could we just default to that if one isn't set?

Reply:
The default server is already set if the user didn't explicitly specify a server: https://github.com/clementine-player/Clementine/pull/6920/files#diff-171d1bdc283ca20217b83a3e9b3f86fbf8acfbae60660180920325b8afc891e0R50
Do you mean that if the current server is not valid, it should be reset automatically to the default one?

Reply:
Assuming most users will never set a different server, you could consider falling back to the default in ReloadSettings and skipping the settings popup on first use.

Reply:
> A nice follow-up feature would be an option to add a station to the "Your radio streams" list.

Added in https://github.com/clementine-player/Clementine/pull/6920/commits/22c02981452b6e6128fdae0f3d34eaf90a772e33

Reply:
> Assuming most users will never set a different server, you could consider falling back to the default in ReloadSettings and skipping the settings popup on first use.

I'm not sue i get what you mean. On first use the QSetting value will be empty and use the default defined in [RadioBrowserSettingsPage::Load](https://github.com/clementine-player/Clementine/pull/6920/files#diff-171d1bdc283ca20217b83a3e9b3f86fbf8acfbae60660180920325b8afc891e0R50)
The user must explicitly change it in order to remove the default value.

Reply:
Right now, when I click on the icon for the first time, it opens the settings. I'm just saying that that's not absolutely necessary for use of the feature. But ultimately, the user flow is your choice.

One small annoyance, though: Click the icon, the settings comes up, hit cancel. The next time you click, nothing happens. Not a huge deal since you an still go to settings to configure at that point.

Reply:
Actually, it looks like I have to restart clementine to get it working again.

Reply:
I finally got what you mean, i reworked the way the default server is set at the first startup and added some more error handling